### PR TITLE
PR for Issue #165

### PR DESCRIPTION
**Pull request by mincong-h**
- Issue #165: Use `Projections#id()` instead of `DocumentBuilderIndexedEntity#getIdPropertyName()` to avoid null reference issue for provided IDs.
- Issue #168: Test the batch job runnability over a poorly-formed entity

### CODE REVIEW

**Comment by mincong-h** [comment]
Could you review this, @yrodiere? Do I need to add some unit tests to prove the ID is not null in any case?

File: core/src/main/java/org/hibernate/search/jsr352/internal/steps/lucene/EntityReader.java [hunk 1]
```diff
@@ -242,7 +242,8 @@ private ScrollableResults buildScrollUsingHQL(StatelessSession ss, String HQL) {
 	private ScrollableResults buildScrollUsingCriteria(StatelessSession ss,
 			PartitionBound unit, Object checkpointId, JobContextData jobData) {
 
-		String idName = MassIndexerUtil.getIdName( entityType, session );
+		String idName = Projections.id().toString();
```
Comment (yrodiere) [inline core/src/main/java/org/hibernate/search/jsr352/internal/steps/lucene/EntityReader.java]: Using `toString()` for anything else than debugging seems like a bad idea to me in general, unless an interface/class documents a specific behavior for this method. In this case, it seems that `Projection` does not give any special meaning to `toString()`. It seems the id projection's `toString()` method will just return "id".

Judging from the code in `Projections` and `org.hibernate.criterion.Order.toSqlString(Criteria, CriteriaQuery)`, I don't think this will work as soon as the id column is named something other than `id`. I'm not 100% sure, but this probably should be checked. Maybe you could test that?

From what I can see in `org.hibernate.criterion.Order.toSqlString(Criteria, CriteriaQuery)`, there is some support for aliases, though. So maybe you could just alias the id projection to something (say, `Projections.alias( Projections.id(), "aliased_id" )`) and order by the alias (`Order.property( "aliased_id" )`)?

**Review by yrodiere — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/hibernate/search/jsr352/internal/steps/lucene/EntityReader.java (hunk 1 above)
Comment (mincong-h) [reply core/src/main/java/org/hibernate/search/jsr352/internal/steps/lucene/EntityReader.java]: > Using toString() for anything else than debugging seems like a bad idea to me in general ...

I don't have other choice than using the `toString()` method. You suggested me to use `Projections#id()`, but it returns an `IdentifierProjection`, which is not usable for creating an object `org.hibernate.criterion.Order`. All constructors of `Order` requires a String as argument, `IdentifierProjection` is not a String so it doesn't fit any constructor.

> ... Maybe you could test that?

I don't understand the 2nd paragraph. Do you want me to create an entity type `MyEntity.class` which has a customized name of the identifier and test it through the batch indexing job?

```java
@Entity
@Indexed
public class MyEntity implements Serializable {

    @Id
    @DocumentId
    private int customId;  // sth like this: 
                           // id column is named sth other than id?

    ...
}
```

As for the 3rd paragraph,

- If the Alias were created using method `Projections#alias(Project, String)` as you suggested, the `toString()` method can be avoid. But how to use this alias? Order by aliased projection does not exists. No matter we're constructing an ascending order `org.hibernate.criterion.Order#asc(String)` or a descending order `org.hibernate.criterion.Order#desc(String)`, both of them use String as an argument and not an aliased projection. As for `Order.property( "aliased_id" )`, this method `Order#property(String)` doesn't exist. (I'm using Hibernate ORM _5.1.0.Final_).
- If the Alias were created using method `org.hibernate.Criteria#createAlias(String, String)`, the `toString()` method is needed, because it requires a String argument.

Comment (mincong-h) [reply core/src/main/java/org/hibernate/search/jsr352/internal/steps/lucene/EntityReader.java]: So basically, I don't see how to use Alias and I don't know how to change the code. Could you give me more detail, ideally some lines of code?

Comment (yrodiere) [reply core/src/main/java/org/hibernate/search/jsr352/internal/steps/lucene/EntityReader.java]: ### 1. Testing

> Do you want me to create an entity type MyEntity.class which has a customized name of the identifier and test it through the batch indexing job?

Yes, that's what I meant. I think the batch won't work if the id property is named `customId`, and that's what this issue is about.

Keep in mind that you could also have a non-id property named "id", while the actual id property (annotated with `@Id`) would be named `customId`:

```
@Entity
@Indexed
public class MyEntity implements Serializable {

    @Id
    @DocumentId
    private int customId; // id column is named sth other than id

    @Basic
    @Column(name = "id")
    private String id;  // property named "id", but is not the id!

    ...
}
```

I wouldn't recommend this for real-world applications, but it's entirely possible, and it's something we should address.

### 2. Ordering with an aliased projection on the ID

> As for Order.property( "aliased_id" ), this method Order#property(String) doesn't exist. (I'm using Hibernate ORM 5.1.0.Final).

Right, the method was `Order.asc`, not `Order.property`.

> So basically, I don't see how to use Alias and I don't know how to change the code. Could you give me more detail, ideally some lines of code?

The purpose of an alias is to define a string that can later be used to refer to the projection that was aliased. So the fact that  `Order.asc` requires a string is not a problem: it's just what aliases are here for.

If you only had to project on IDs, the code would be like this (excluding the various other calls like `scrollMode` etc.):

```
return criteria.setProjection( Projections.alias( Projections.id(), "aliased_id" ) )
   .addOrder( Order.asc( "aliased_id" ) );
```

So I suppose the issue was that you don't want the result of the query to be the ID, but the whole entity. So you can't simply use `criteria.setProjection( Projections.alias( Projections.id(), "aliased_id" ) )` . And setting up a projection on both the ID and the whole entity indeed seems a bit hard. Judging from the implementation code, I'm not even sure it's possible...

So, plan B: retrieve the ID property name through Hibernate ORM's metadata: `sessionFactory.getClassMetadata( entityType ).getIdentifierPropertyName()`. And put *that* in `idName`.
Should have been plan A, really: it's much simpler. Sorry about that. I get carried away, sometimes :)

By the way, don't forget that `MassIndexerUtil.getIdName` is also used elsewhere. In `org.hibernate.search.jsr352.internal.steps.lucene.PartitionMapper.buildScrollableResults(StatelessSession, Session, Class<?>, Set<Criterion>)` in particular. The issue should be fixed there, too. You could have used aliases there, since you're only projecting on the ID, but sticking to the same solution seems a better idea.

### 3. Side note: `Object.toString()`
To be clear, the issue with using `toString` isn't that it does not suit my taste: it is that the method does not return the string you need. It's like picking a random getter in the class just because it returns a string. From the javadoc of `java.lang.Object#toString`:

> Returns a string representation of the object. In general, the
> `toString` method returns a string that
> "textually represents" this object. The result should
> be a concise but informative representation that is easy for a
> person to read.

... so the intended purpose of this method is to format the object for *humans*, not to serialize it for other parts of the application. In some case it may do both, but `IdentifierProjection.toString` clearly does not.

Comment (mincong-h) [reply core/src/main/java/org/hibernate/search/jsr352/internal/steps/lucene/EntityReader.java]: After your very detailed explanation, I understand much better now! I'll avoid using `toString()` method in the future. Actually, the method Java code with aliased projection is similar to the `AS` syntax in SQL. So the following block

```java
ScrollableResults scroll = criteria
        .setProjection( Projections.alias( Projections.id(), "aliasedId" ) )
        .addOrder( Order.asc( "aliasedId" ) )
        ...
```

is similar to SQL

```sql
SELECT id_column AS aliasedId
FROM my_table
ORDER BY aliasedId
```

Btw, I'm not sure whether `aliasedId` or `aliased_id` is better. The 1st one in camel case is more Java-like and the 2nd one in snake case is more SQL-like. If you prefer snake case, please let me know.

`Projections.alias( Projections.id(), "aliasedId" )` is great, however, it does work only for ID projection. So in the class `org.hibernate.search.jsr352.internal.steps.lucene.EntityReader`, I need to use you plan B solution to avoid projection issue: what we want to obtain is entity itself with all the fields. So I changed to code to:

```java
String idName = sessionFactory.getClassMetadata( entityType )
        .getIdentifierPropertyName();
```

> By the way, don't forget that `MassIndexerUtil.getIdName` is also used elsewhere.

Method deleted in commit 15d2952. Thanks for you hint! 👍 

> You could have used aliases there, since you're only projecting on the ID, but sticking to the same solution seems a better idea.

No, I can't use the projection. The scrolling result in `EntityReader` isn't a projection on ID, it loads the entire entity (we may change this logic as you mentioned in issue #164, but let's discuss it later). Therefore,  I need to use your plan B as shown in previous paragraph.

**Review by yrodiere — APPROVED** [review]
Looks good to me! Thanks :)
Can I let you do the merge? I'd personally squash most of the commits and rebase in this particular case, but I don't know what you usually do in this repo, so...

**Comment by mincong-h** [comment]
I didn't know much about squash and rebase yet, since almost all of my previous projects have only one contributor (me 😢 ), and I simply pushed to master... Maybe I can learn more when we integrate this project into Hibernate Search's repo. I'll merge this PR on GitHub for instance.
